rust/lance-table/src/utils/stream.rs

```diff
@@ -0,0 +1,518 @@
+    FutureExt, Stream, StreamExt,
+};
+use lance_arrow::RecordBatchExt;
+use lance_core::{
+    utils::{address::RowAddress, deletion::DeletionVector},
+    Result, ROW_ID, ROW_ID_FIELD,
+};
+use lance_io::ReadBatchParams;
+
+type BatchFut = BoxFuture<'static, Result<RecordBatch>>;
+/// A task, emitted by a file reader, that will produce a batch (of the
+/// given size)
+pub struct BatchTask {
+    task: BatchFut,
+    num_rows: u32,
+}
+type BatchTaskStream = BoxStream<'static, BatchTask>;
+type BatchFutStream = BoxStream<'static, BatchFut>;
+
+struct MergeStream {
+    streams: Vec<BatchTaskStream>,
+    next_batch: FuturesOrdered<BatchFut>,
+    next_num_rows: u32,
+    index: usize,
+}
+
+impl MergeStream {
+    fn emit(&mut self) -> BatchTask {
+        let mut iter = std::mem::take(&mut self.next_batch);
+        let task = async move {
+            let mut batch = iter.next().await.unwrap()?;
+            while let Some(next) = iter.next().await {
+                let next = next?;
+                batch = batch.merge(&next)?;
+            }
+            Ok(batch)
+        }
+        .boxed();
+        self.next_num_rows = 0;
+        BatchTask {
+            task,
+            num_rows: self.next_num_rows,
+        }
+    }
+}
+
+impl Stream for MergeStream {
+    type Item = BatchTask;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        loop {
+            let index = self.index;
+            match self.streams[index].poll_next_unpin(cx) {
+                std::task::Poll::Ready(Some(batch_task)) => {
+                    if self.index == 0 {
+                        self.next_num_rows = batch_task.num_rows;
+                    } else {
+                        debug_assert_eq!(self.next_num_rows, batch_task.num_rows);
+                    }
+                    self.next_batch.push_back(batch_task.task);
+                    self.index += 1;
+                    if self.index == self.streams.len() {
+                        self.index = 0;
+                        let next_batch = self.emit();
+                        return std::task::Poll::Ready(Some(next_batch));
+                    }
+                }
+                std::task::Poll::Ready(None) => {
+                    return std::task::Poll::Ready(None);
+                }
+                std::task::Poll::Pending => {
+                    return std::task::Poll::Pending;
+                }
+            }
+        }
+    }
+}
+
+/// Given multiple streams of batch tasks, merge them into a single stream
+///
+/// This pulls one batch from each stream and then combines the columns from
+/// all of the batches into a single batch.  The order of the batches in the
+/// streams is maintained and the merged batch columns will be in order from
+/// first to last stream.
+///
+/// This stream ends as soon as any of the input streams ends (we do not
```

Comment:
It requires each stream is equal sized? What was the use case for this?
It seems oddly restricted to me without knowing much context of its usage

Reply:
In legacy format, we require batch_size to be exactly equal. In new format, there are no batches, so we can read with an arbitrary batch size. Therefore, I don't think this is a challenging limitation at all.

Reply:
Correct.  We can lift the limitation if needed by slicing and dicing the various batches, it just adds a lot of complexity where none is needed.